File: Src/Compilers/CSharp/Source/Syntax/SyntaxList.WithManyChildren.cs
// Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using System;

namespace Microsoft.CodeAnalysis.CSharp.Syntax
{
    internal partial class SyntaxList
    {
        internal class WithManyChildren : SyntaxList
        {
            private readonly ArrayElement<SyntaxNode>[] children;

            internal WithManyChildren(Syntax.InternalSyntax.SyntaxList green, SyntaxNode parent, int position)
                : base(green, parent, position)
            {
                this.children = new ArrayElement<SyntaxNode>[green.SlotCount];
            }

            internal override SyntaxNode GetNodeSlot(int index)
            {
                return this.GetRedElement(ref this.children[index].Value, index);
            }

            internal override SyntaxNode GetCachedSlot(int index)
            {
                return this.children[index];
            }

            public override TResult Accept<TResult>(CSharpSyntaxVisitor<TResult> visitor)
            {
                throw new NotImplementedException();
            }

            public override void Accept(CSharpSyntaxVisitor visitor)
            {
                throw new NotImplementedException();
            }
        }
    }
}

File: src/Compilers/VisualBasic/Portable/Syntax/SyntaxList.WithManyChildren.vb
' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax

    Friend Partial Class SyntaxList

        Friend Class WithManyChildren
            Inherits SyntaxList

            Private ReadOnly _children As ArrayElement(Of SyntaxNode)()

            Friend Sub New(green As InternalSyntax.SyntaxList, parent As SyntaxNode, position As Integer)
                MyBase.New(green, parent, position)
                _children = New ArrayElement(Of SyntaxNode)(green.SlotCount - 1) {}
            End Sub

            Friend Overrides Function GetNodeSlot(index As Integer) As SyntaxNode
                Return GetRedElement(_children(index).Value, index)
            End Function

            Friend Overrides Function GetCachedSlot(i As Integer) As SyntaxNode
                Return TryCast(_children(i).Value, VisualBasicSyntaxNode)
            End Function
        End Class
    End Class
End Namespace
